### PR TITLE
Feature : Code action Import all missing symbols

### PR DESCRIPTION
**Pull request by coline-lunatech**
Add the code action "import all missing symbols that are unambiguous" for issue #1610

### CODE REVIEW

File: tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala [hunk 1]
```diff
@@ -78,26 +78,102 @@ class ImportMissingSymbolLspSuite
...
+        |${CreateNewSymbol.title("Future")}
+        |${CreateNewSymbol.title("Instant")}
+        |${CreateNewSymbol.title("ListBuffer")}
+        |${CreateNewSymbol.title("Future")}
+        |${CreateNewSymbol.title("ListBuffer")}
+        |""".stripMargin,
+    """|package a
+       |
+       |import java.time.Instant
+       |
```
Comment (coline-lunatech) [inline tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala:170]: I don't know why there's a space here ... Moreover tests sometimes pass and sometimes not cause the imports are reversed

File: metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala [hunk 2]
```diff
@@ -43,14 +44,42 @@ class ImportMissingSymbol(compilers: Compilers) extends CodeAction {
         }
     }
 
+    def addImportAllMissingSymbol(
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala:47]:
```suggestion
    def importMissingSymbols(
```
Maybe this to better align with the other `importMissingSymbol` in the file.

File: metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala [hunk 3]
```diff
@@ -43,14 +44,42 @@ class ImportMissingSymbol(compilers: Compilers) extends CodeAction {
...
+        codeActions: Buffer[l.CodeAction]
+    ): Seq[l.CodeAction] = {
+      val uniqueCodeActions = codeActions
+        .groupBy(_.getDiagnostics())
+        .values
+        .filter(_.length <= 1)
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala:53]: What would be a scenario where we'd want `< 1`? Aren't we always targeting when there is `1` exactly?

File: metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala [hunk 4]
```diff
@@ -43,14 +44,42 @@ class ImportMissingSymbol(compilers: Compilers) extends CodeAction {
         }
     }
 
+    def addImportAllMissingSymbol(
+        codeActions: Buffer[l.CodeAction]
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala:48]: You can probably just leave this as a `Seq` instead of a buffer. Down below at the end of the file where this is `flattened` and passed it, it can just stay a `Seq`. Then in this method, you could change the logic a bit that instead of the `if` statement adding something to the buffer, you could do something like


```scala
val codeActionsToReturn = If(uniqueCodeActions.length > 1) {
  ... // create new code action
  codeActions :+ newCodeAction
} else {
  codeActions
}
codeActionsToReturn
```

File: metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala [hunk 5]
```diff
@@ -60,4 +89,6 @@ object ImportMissingSymbol {
   def title(name: String, packageName: String): String =
     s"Import '$name' from package '$packageName'"
 
+  def allSymbolsTitle: String =
+    s"Import all missing symbols that are unambiguous"
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala:93]: I'm curious if showing the `unambiguous` is a bit verbose to the user.
```suggestion
    s"Import all possible missing symbols"
```
What do you think?

File: tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala (hunk 1 above)
Comment (ckipp01) [reply tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala:170]: If you add a sort in the code action that returns them, does it fix this?

File: metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala (hunk 3 above)
Comment (coline-lunatech) [reply metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala:53]: yep that's true

File: tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala (hunk 1 above)
Comment (coline-lunatech) [reply tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala:170]: I wanted to keep the "import all" above all the others but it does not work quite well anyway.. I'll try different ways

Comment (coline-lunatech) [reply tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala:170]: It will not fix anyway because it's not the order of actions but the order of TextEdits that could matter.
But actually it looks like it's kind of random, text edits order is always the same but they seems not to be applied in that order.

=> after some checks I found that it's only on tests. On files when I use the codeAction they are always applied in the same order. (Could be the textEdits.scala file)

Comment (coline-lunatech) [reply tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala:170]: About the spaces it's also a textEdits issue. If a file has no imports yet, imports textEdits are prepend with a \n. In our case we concatenate textEdits so they all are prepend with it..

I'm not sure I can fix this two issues

Comment (ckipp01) [reply tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala:170]: Ahhhh of course...
https://github.com/scalameta/metals/pull/1051
I should have known that.. seeing that I added it 😬 .

I suppose this is an edge case that you'll only hit on if you are triggering this code action from a file that has no imports yet. 🤔 I'm trying to think what the best approach to fixing this is. We could just check the text edits and strip out the leading `\n` on the second textEdit, but that will still fail in this test if it's flaky. I'm not 100% sure why the textEdits aren't applied in a uniform way in tests. @tgodzik may know more on that.

Comment (tgodzik) [reply tests/unit/src/test/scala/tests/codeactions/ImportMissingSymbolLspSuite.scala:170]: We do sort them by start to apply them properly:
```
      val positions = edits
        .map(edit => edit -> edit.getRange.toMeta(input))
        .sortBy(_._2.start)
```
in `TextEdits.scala`. I am not sure what will be the behaviour in case of multiple imports, but I believe we need to fix it there.

File: metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala [hunk 6]
```diff
@@ -43,14 +43,42 @@ class ImportMissingSymbol(compilers: Compilers) extends CodeAction {
...
+          .distinct
+
+        allSymbols.setTitle(ImportMissingSymbol.allSymbolsTitle)
+        allSymbols.setKind(l.CodeActionKind.QuickFix)
+        allSymbols.setDiagnostics(diags.asJava)
+        allSymbols.setEdit(new l.WorkspaceEdit(Map(uri -> edits.asJava).asJava))
+
+        allSymbols +: codeActions
+      } else { codeActions }
+      allCodeActions
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala:72]: This could be dropped. Just have the if-else as the last expression.

File: metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala [hunk 7]
```diff
@@ -43,14 +43,42 @@ class ImportMissingSymbol(compilers: Compilers) extends CodeAction {
...
+        val uri = params.getTextDocument().getUri()
+        val diags = uniqueCodeActions.flatMap(_.getDiagnostics().asScala)
+        val edits = uniqueCodeActions
+          .flatMap(_.getEdit().getChanges().get(uri).asScala)
+          .distinct
+
+        allSymbols.setTitle(ImportMissingSymbol.allSymbolsTitle)
+        allSymbols.setKind(l.CodeActionKind.QuickFix)
+        allSymbols.setDiagnostics(diags.asJava)
+        allSymbols.setEdit(new l.WorkspaceEdit(Map(uri -> edits.asJava).asJava))
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala:68]: I think we could just join all the textEdits into a single text edit, which might help with the tests and the additional newline.

Comment (coline-lunatech) [reply metals/src/main/scala/scala/meta/internal/metals/codeactions/ImportMissingSymbol.scala:75]: I managed to fix everything that way without changing anything to the TextEdits.scala file. I don't know if this is the right way but it seems that imports are always coming in different order so maybe it is.

**Comment by ckipp01** [comment]
Merging since the testing failure is unrelated.

```
[error] 	tests.debug.BreakpointDapSuite
```

**Comment by ckipp01** [comment]
O, and one other nitpick to remember for the future @coline-lunatech, try to keep your commits a bit more concise. We don't have any strict rules on single commits vs multiple commits, but it's always a good idea to have your commits be able to stand alone and be descriptive. Commits like `more styling` can probably just be rolled into another commit to not make the commits to noisy.
